vms/platformvm/state/subnet_only_validator.go

```diff
@@ -166,3 +248,184 @@ func deleteSubnetOnlyValidator(
+		// subnetID+nodeID as this L1 validator.
+		_, err := state.GetCurrentValidator(sov.SubnetID, sov.NodeID)
+		if err == nil {
+			return ErrConflictingSubnetOnlyValidator
+		}
+		if err != database.ErrNotFound {
+			return err
+		}
+
+		has, err := state.HasSubnetOnlyValidator(sov.SubnetID, sov.NodeID)
+		if err != nil {
+			return err
+		}
+		if has {
+			return ErrDuplicateSubnetOnlyValidator
+		}
+	default:
+		return err
+	}
+
+	if prevWeight != sov.Weight {
+		weight, err := state.WeightOfSubnetOnlyValidators(sov.SubnetID)
+		if err != nil {
+			return err
+		}
+
+		weight, err = math.Sub(weight, prevWeight)
+		if err != nil {
+			return err
+		}
+		weight, err = math.Add(weight, sov.Weight)
+		if err != nil {
+			return err
+		}
+
+		d.modifiedTotalWeight[sov.SubnetID] = weight
+	}
+
+	switch {
+	case prevActive && !newActive:
+		d.netAddedActive--
+	case !prevActive && newActive:
+		d.netAddedActive++
+	}
+
+	if prevSOV, ok := d.modified[sov.ValidationID]; ok {
+		d.active.Delete(prevSOV)
+	}
+	d.modified[sov.ValidationID] = sov
+
+	subnetIDNodeID := subnetIDNodeID{
+		subnetID: sov.SubnetID,
+		nodeID:   sov.NodeID,
+	}
+	d.modifiedHasNodeIDs[subnetIDNodeID] = !sov.isDeleted()
+	if sov.isActive() {
+		d.active.ReplaceOrInsert(sov)
+	}
+	return nil
+}
+
+type activeSubnetOnlyValidators struct {
+	lookup map[ids.ID]SubnetOnlyValidator
+	tree   *btree.BTreeG[SubnetOnlyValidator]
+}
+
+func newActiveSubnetOnlyValidators() *activeSubnetOnlyValidators {
+	return &activeSubnetOnlyValidators{
+		lookup: make(map[ids.ID]SubnetOnlyValidator),
+		tree:   btree.NewG(defaultTreeDegree, SubnetOnlyValidator.Less),
+	}
+}
+
+func (a *activeSubnetOnlyValidators) get(validationID ids.ID) (SubnetOnlyValidator, bool) {
+	sov, ok := a.lookup[validationID]
+	return sov, ok
+}
+
+func (a *activeSubnetOnlyValidators) put(sov SubnetOnlyValidator) {
+	a.lookup[sov.ValidationID] = sov
+	a.tree.ReplaceOrInsert(sov)
+}
+
+func (a *activeSubnetOnlyValidators) delete(validationID ids.ID) bool {
+	sov, ok := a.lookup[validationID]
+	if !ok {
+		return false
+	}
+
+	delete(a.lookup, validationID)
+	a.tree.Delete(sov)
+	return true
+}
+
+func (a *activeSubnetOnlyValidators) len() int {
+	return len(a.lookup)
+}
+
+func (a *activeSubnetOnlyValidators) newIterator() iterator.Iterator[SubnetOnlyValidator] {
+	return iterator.FromTree(a.tree)
+}
+
+func (a *activeSubnetOnlyValidators) addStakersToValidatorManager(vdrs validators.Manager) error {
+	for validationID, sov := range a.lookup {
+		pk := bls.PublicKeyFromValidUncompressedBytes(sov.PublicKey)
+		if err := vdrs.AddStaker(sov.SubnetID, sov.NodeID, pk, validationID, sov.Weight); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func addSoVToValidatorManager(vdrs validators.Manager, sov SubnetOnlyValidator) error {
+	nodeID := sov.effectiveNodeID()
+	if vdrs.GetWeight(sov.SubnetID, nodeID) != 0 {
+		return vdrs.AddWeight(sov.SubnetID, nodeID, sov.Weight)
```

Comment:
This can occur if the `effectiveNodeID` is the `ids.EmptyNodeID` (it does not always happen if that is the case, but it can.

This function is only called if there is no prior version of this SoV.

I added a test that fails if the weight is not increased as expected.